### ISSUE TITLE
Authentication filters

### ISSUE THREAD

**Issue by Charlotte-Holt** [OpenLiberty/docs#939]
From LC: Advanced topic - e.g., In a big company, there are different authentication for customers (e.g., social login/OIDC) vs employees (e.g., SPNEGO). It can also specify only certain social logins are available on specific endpoints (links back to using social login). 

Provide an introduction to what authentication filters are and why you'd use them, then provide the example config from the 'examples' section of the KC topic. And a link to the Social Login gen topic (and any of the other features that can use auth filters).

SPNEGO config examples w/ #954 

**Scenario 1**: This topic needs to be written for a developer writing cloud-native apps that will be containerized and deployed to the cloud.

**Resources**:
- https://www.ibm.com/support/knowledgecenter/en/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/rwlp_auth_filter.html but don't include the list of elements in the 'Supported elements' section (that info should be in the generated doc).

**Comment by dmuelle** [comment]
@lauracowen draft available for review at https://draft-openlibertyio.mybluemix.net/docs/ref/general/#auth-filter.html

**Comment by dmuelle** [comment]
SME reviewed by Ut Le

**Comment by chirp1** [comment]
@dmuelle Hi David, very nicely written!

Here are a few comments:

- Always have at least one sentence between two headers. So, add at least one meaningful sentence between "Authentication filter examples" and "Request URL contains a pattern".
- Keep your headers consistent. The "Excluding patterns" and "Using all sub-elements" headers are not consistent with the rest of the headers. Redo the headers, following the pattern of the other headers. 
- Do you need so much information in "Excluding Patterns"? Couldn't you just have the statement with notcontain in the example? If the other info is required, start the paragraph talking about notcontain since that is the point of the heading. If the other info is not required, but you want to discuss it, create another example for it. 
- For "Using all sub-elements", the "Target application is named myApp" is not following the pattern of the bullets above it. Rewrite the sentence to follow the pattern.

**Comment by dmuelle** [comment]
@chirp1 comments addressed in OpenLiberty/docs#1110 . Will rebuild the draft site to pick up changes this afternoon

**Comment by lauracowen** [comment]
- "In this case, a web application can be secured with the Social Media Login feature, while an associated microservice on the same server is secured with the MicroProfile JSON Web Token feature." - I'm not sure about the wording. It's not the features per se that are securing it, it's the technologies. So the web application can be secured with a social media login, or the web application can be secured by a social media provider. I'm not sure you would secure a microservice with the MP JWT exactly as an alternative. Check this but I think a better example might be using LDAP vs social media login? Or something like that? The subsequent paragraph does this and reads better, I think.
- The sentence about where the auth mechanisms in OL are managed is fine (though "feature configuration" should be "server configuration"). But don't provide instructions here on how to do it. Instead, talk about it conceptually - eg it's configured in the server configuration (server.xml) file, where you specify each auth mechanism and the URL patterns for which that auth mechanism should be used. For example, the social media login form might be presented for URLs ending in `/xxxxx`, while the end-user is prompted for their company (LDAP) login details for URLs ending in `/yyyy`. Ask Bruce or someone for more plausible URL examples. Just focus on getting across how useful it is here. Maybe a simple diagram too - it doesn't have to be clever, just so that the concept comes a cross at a glance.
- I don't think we should list the _features_ that support auth filters. Instead, list the authentication mechanisms (from the user perspective) ie (from that list): SAML, SPNEGO, social media providers, JSON Web Tokens (JWT). Features are just a means to an end - not what the developer really cares about - they just enable the features they need to get the tech/mechanisms they want to use (or have to support in their app). Then link to the relevant concept topics that we're writing about those mechanisms. Don't mention the ones we're not writing concept topics for (eg OpenID) as we're focused on the auth mechanisms we've got topics for. The autogen topics of course show that the OpenID feature, for instance, does support authfilters and that's fine if they need to know that.
- Can you move the config examples to the autogen feature doc (create a new issue for it so we can track it separately) then link to that when it's done. Now, the trick here is how to do this if it's relevant to multiple different features...It may be that it would work as an include to an include (ie create the include files for each of the features [SAML, SPENGO, social media login, MP JWT] then put include statements in each of them that pulls in a single file that contains the auth filter example config. But reduce the number of examples down because there are too many there (I like that there's a link off to the Auth Filter config element doc for a full list of elements). Find out which are the commonly most useful configurations and include them only. eg I think the `matchType="notContain"` example is kind of obvious if you've looked at the previous "contain" examples and then look at the full list of config elements available.
This will leave a short-ish topic but I think that's fine.

**Comment by dmuelle** [comment]
- "In this case, a web application can be secured with the Social Media Login feature, while an associated microservice on the same server is secured with the MicroProfile JSON Web Token feature." - I'm not sure about the wording. It's not the features per se that are securing it, it's the technologies. **revised**
- The sentence about where the auth mechanisms in OL are managed is fine (though "feature configuration" should be "server configuration"). But don't provide instructions here on how to do it. Instead, talk about it conceptually - **revised**
- I don't think we should list the _features_ that support auth filters. Instead, list the authentication mechanisms (from the user perspective) ie (from that list): SAML, SPNEGO, social media providers, JSON Web Tokens (JWT). **revised** 
- Can you move the config examples to the autogen feature doc (create a new issue for it so we can track it separately) then link to that when it's done. Now, the trick here is how to do this if it's relevant to multiple different features... ** How best to do this? Putting copies of the file in each of the feature directories or creating a new `authfilter` directory w/ the file in it and including to the features from there?

**Comment by ManasiGandhi** [comment]
@dmuelle The topic looks good! A small suggestion. 

“You can use authentication filters to specify which authentication mechanism Open Liberty uses to process different HTTP servlet requests.” 
to
 “You can use authentication filters to specify the authentication mechanism for Open Liberty to process different HTTP servlet requests.”

**Comment by lauracowen** [comment]
Could you remove the OpenID feature config from this topic now (as it's not one that's a priority to promote), and then move the links to the example configurations in that part of the topic instead of in See Also? Actually, if you want a piece of code example actually showing here, you could tag the example in the social media login feature and 'include' it in here. Then put the links to the other two pages beneath it as 'other examples'.

BTW, the Social Media Login link gives me a 404 in the auth filter topic.

**Comment by dmuelle** [comment]
Replaced OpenID example with Social Media Login example and moved the other examples out of "see also" and into the topic body. However- as this is now a pretty short topic that only meant moving them up one line :) 

See what you think

https://draft-openlibertyio.mybluemix.net/docs/ref/general/#authentication-filters-specifying-mechanism.html

**Comment by lauracowen** [comment]
Thanks

Could you mention in the blurb before the example that it's from the Social Media Login feature and link to it? (so they can find that in context like they can for the other features)

Also, did you look at whether it's possible to just use an asciidoc tags and 'include' statement to pull the example directly from example.adoc file? (rather than copy/paste of the same example in both locations)

**Comment by dmuelle** [comment]
Pulled in the example and added a link to Social Login

**Comment by dmuelle** [comment]
Hi Laura- after a bit of trial and error, I was able pull in the socialLogin example to the authfilter docs using tags. However- I'm not sure of the best way to get the updated Social Media Login 1.0 example file over to master when the time comes. 

Since the examples are only on drafty currently, I didn't have a local copy of the file and had to open a new branch, based on draft, in order to edit it. But I can't PR that branch to dev (bcse it's draft-based).

I looked up the PR where the examples were originally added to draft to find a branch where I could make the edits but it looks like they came directly from your fork of master: https://github.com/OpenLiberty/docs/pull/1035
Just wanted to make a note oof this so we don't forget before this doc goes to dev

**Comment by dmuelle** [comment]
Resolved- updated examples file is now on dev

**Comment by utle** [comment]
- I am not sure what is the Limit SPNEGO authentication?

`Simple and Protected GSSAPI Negotiation Mechanism (SPNEGO): Limit SPNEGO authentication`

- Can we improve this comments?

`<!-- only requests with "/mywebapp" in them will be authenticated using Google. -->`

maybe ?

`
 `

**Comment by dmuelle** [comment]
@utle Thanks for reviewing- I have updated the doc

- SPNEGO link- I changed the subheading to match the updated title of the relevant linked example: `Simple and Protected GSSAPI Negotiation Mechanism (SPNEGO): Apply SPNEGO authentication to a subset of HTTP requests`

- code comment- I've removed the comment since the introductory sentence that precedes the code block provides the necessary information. I updated the explanation to reflect your suggestion above. 

https://draft-openlibertyio.mybluemix.net/docs/ref/general/#authentication-filters-specifying-mechanism.html

Let me know if this topic needs further edits- thanks!

**Comment by utle** [comment]
developer approved

**Comment by dmuelle** [comment]
New draft URL: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.11/authentication-filters.html

**Comment by dmuelle** [comment]
changes made in new editing pass:
- pattern->URL pattern, filter-> authentication filter per DQTI ch6 Use technical terms consistently and appropriately
- reorganized/re-chunked information to improve topic coherency per DQTI ch6 WRITE COHERENTLY
- opened a docbug to fix the code comment that Ut mentions in his review- since the example is now an include, the changes I previously made were lost. Making the edit in the Social Login feature example doc, from which the example here is included

**Comment by dmuelle** [comment]
comment from @utle in #2761 

We need to pick the update of authFilter metatype that have more description and samples.

**Comment by dmuelle** [comment]
@utle when was the metatype.properties updated? If prior to 20.0.0.11, it should be in the build already. We pull the gendoc with each 4 week release. Need to investigate if we are missing content. 

@cbridgha - do we need to take any additional steps to get updates for metatype.properties files into the OL gendoc? @utle mentions here and w/r/t `ssl` that there are updates to the properties that are not showing up in the docs.

**Comment by chirp1** [comment]
@dmuelle Hi David, 
Here are my comments:
- The links to the JWT, SPNEGO, and Social media login features each have an example for authentication filters. However, the SAML feature does not. Add an example authentication filter to the SAML feature. Also, ensure that all four features link to the Authentication filter references. Only two of them do.
- The topic that you drew from https://www.ibm.com/support/knowledgecenter/en/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/rwlp_auth_filter.html mentions userAgent, host, webApp, remoteAddress, and requestUrl elements.  and has at least one example for each. Ensure that among JWT, SPNEGO, Social media login, and SAML topics that you cover these elements. You don't have to have examples of all five elements in each feature, but you should appropriately cover the five elements among them. For instance, the remoteAddress, and requestUrl. elements are not in an authentication filter example in any of the four features, but likely should be added. If you think that more than one example is needed for each feature to cover the examples pertinent for customers, then add additional examples as appropriate to the features.
- I also think that you should mention the five elements in your topic and I think that you should move the "Authentication filter" link from "See also" into the body of your topic since the five elements are at the Authentication filter link. 
- You could also create a list or create a table that indicates the five elements and then link to the examples in the four social media login features, so that the users don't have to hunt for the examples.. 
- Rework the sentence that has "including the following options: " so that you mention that they are features instead of options. 
- You might get rid of the Social media login example in your topic as it seems a bit redundant since each of the four features should have authentication filter examples in them.
- For the "Simple and Protected GSSAPI Negotiation...." and "MicroProfile JSON Web Token " links, they are only linked to the topic, even though the text seems to indicate that each should be linking to a section title in each topic.
- The two sentences "In the server.xml file, you can specify authentication filters to control which authentication mechanism secures certain types of requests. In an application where different kinds of protected resources require different types of authentication, authentication filters can control how authentication is managed for each resource." seem a bit vague and sort of redundant. Crisp up the text.

**Comment by dmuelle** [comment]
hi @chirp1 thanks for reviewing- I made the following changes per your review

- [ ] The links to the JWT, SPNEGO, and Social media login features each have an example for authentication filters. However, the SAML feature does not. Add an example authentication filter to the SAML feature. Also, ensure that all four features link to the Authentication filter references. Only two of them do. **I removed  SAML from the list of examples rather than adding a configuration example for it that had not been otherwise identified as necessary. If we need one, I can reach out to an SME to help to determine what kind of example would be relevant**

- [ ] **Re the config elements mentioned in the KC topic**- I referenced the autogen page in Open Liberty and discovered two additional config elements that were not covered in the KC topic- `cookie` and `requestHeader`. I'm not sure if we should doc all the config elements here, and then have to manually maintain them, or if we should open up a doc bug to improve the descriptions in the autogen. Likewise, if we want to provide examples of each configuration element, we need examples and explanations for the two new elements, which will require more SME input and review. If that is the way we should go I propose holding this topic out from the initial batch of security docs until we can get this info and revise the topic

- [ ] Rework the sentence that has "including the following options: " so that you mention that they are features instead of options. **I left this as is for now, per @lauracowen comment above- "I don't think we should list the features that support auth filters. Instead, list the authentication mechanisms (from the user perspective)"**

- [ ] You might get rid of the Social media login example in your topic as it seems a bit redundant since each of the four features should have authentication filter examples in them. **I've left it in until we determine how we want to doc the other examples- if we decide to include more examples in this topic it should probably be removed**

- [x] For the "Simple and Protected GSSAPI Negotiation...." and "MicroProfile JSON Web Token " links, they are only linked to the topic, even though the text seems to indicate that each should be linking to a section title in each topic.

- [x] The two sentences "In the server.xml file, you can specify authentication filters to control which authentication mechanism secures certain types of requests. In an application where different kinds of protected resources require different types of authentication, authentication filters can control how authentication is managed for each resource." seem a bit vague and sort of redundant. Crisp up the text.

**Comment by dmuelle** [comment]
@chirp1  - I've added a section for `Supported elements` based on the KC topic and added a short example for each element. There is still one element in the autogen not included here as I need more information and it wasn't in the KC topic. I can follow up on that with Ut.

**Comment by dmuelle** [comment]
also- just noticed [@lauracowen comment from way back at the beginning of this issue](https://github.com/OpenLiberty/docs/issues/939#issue-552510871) not to include supported elements because it should be in autogen. I think I agree since the KC topic was already out of date and we didn't know. See what you think about what I've added, I can remove or edit as necessary

**Comment by chirp1** [comment]
@dmuelle Thanks for the explanation about the supported elements being in the autogen. So, yes, remove the supported elements for the authentication filters from your Authentication filers topics. You could, as a possible option  list them all (you are missing the one for cookies) and then link to the autogen authentication filter topic.... Looking over the rest of your updates.

**Comment by chirp1** [comment]
@dmuelle Hi David,
Just the following comments:
- See my comment directly above about supported elements and autogen.
- The link to the section for MicroProfile JSON Web Token still seems to go to the top of the page for me.

**Comment by dmuelle** [comment]
Thanks @chirp1. I removed the supported elements section and just left a link to the authFilter config element. I didn't list the supported elements in the topic since that list would have to be manually maintained and the user can get the same info by clicking through to the autogen. I opened [a docbug to improve the autogen](https://github.com/OpenLiberty/docs/issues/2954).

For the Mp Jwt link- the link is correct but there is a bug with anchor links on the OL site- [there's an issue open to fix it](https://github.com/OpenLiberty/openliberty.io/issues/1886).
